File: python/tilley/method.py
import math
import numpy as np

from helpers import *


def tilley(P, Q, s0, K, sigma, r, T, N, op_type, sharp_boundary=True):
    if op_type not in "CP":
        return
    R = P * Q
    dt = T / N
    discount_rate = math.exp(-r * dt)
    indexes = np.array(list(range(0, R)))
    stocks = np.ndarray(shape=[R, N], dtype=[("val", float), ("current", float)])
    true_indicator = np.zeros(shape=[R, N])
    for path in range(0, R):
        prev = s0
        for step in range(0, N):
            prev = generate_new_sample(prev, dt, sigma, r)
            stocks[path, step]["val"] = prev
        stocks[path, N - 1]["current"] = excercise_option(prev, K, op_type)

    for step in reversed(range(0, N - 1)):
        sort_key = lambda x: stocks[x, step]["val"]
        # Ordenamos en cada paso los valores de la trayectoria de mayor a menor
        indexes = np.array(sorted(indexes, key=sort_key, reverse=(op_type == "P")))
        tentative_indicator = np.full(shape=[R], fill_value=-1)
        holding_values = np.zeros(Q)

        for bundle in range(0, Q):
            bundle_currents = []
            for path in range(0, P):
                path_index = indexes[bundle * P + path]
                current_value = stocks[path_index, step + 1]["current"]
                bundle_currents.append(current_value)
            # En cada agrupamiento se calcula el valor de no ejercer para ese agrupamiento
            holding_value = discount_rate * np.mean(bundle_currents)
            holding_values[bundle] = holding_value
            for path in range(0, P):
                # En base a el valor de no ejercer del agrupamiento se actualiza
                # un indicador tentativo para cada trayectoria de este agrupamiento
                path_index = indexes[bundle * P + path]
                val = stocks[path_index, step]["val"]
                tentative_indicator[bundle * P + path] = (
                    1 if excercise_option(val, K, op_type) > holding_value else 0
                )

        boundary = R
        i = R - 1
        max_zero_len = 0
        while i >= 0:
            j = i
            if tentative_indicator[i] == 1:
                one_len = 0
                while tentative_indicator[j] == 1 and j >= 0:
                    one_len += 1
                    j -= 1
                if max_zero_len < one_len:
                    boundary = j + 1
                i = j
            else:
                zero_len = 0
                while tentative_indicator[j] == 0 and j >= 0:
                    zero_len += 1
                    j -= 1
                max_zero_len = max(zero_len, max_zero_len)
            i = j

        sharp_boundary_indicator = np.zeros(R)
        for i in range(boundary, R):
            sharp_boundary_indicator[i] = 1

        indicator = sharp_boundary_indicator if sharp_boundary else tentative_indicator

        for bundle in range(0, Q):
            holding_value = holding_values[bundle]
            for path in range(0, P):
                path_index = indexes[bundle * P + path]
                val = stocks[path_index, step]["val"]
                stocks[path_index, step]["current"] = (
                    excercise_option(val, K, op_type)
                    if indicator[bundle * P + path] == 1
                    else holding_value
                )

        # El primer 1 del indicador de frontera indica que en ese paso se ejerce
        if len(np.where(indicator == 1)[0]) > 0:
            first_one = np.where(indicator == 1)[0][0]
            true_indicator[indexes[first_one], step] = 1
            for j in range(step + 1, N):
                true_indicator[indexes[first_one], j] = 0

    estimator = 0
    for path in range(0, R):
        step = next(
            (step for step in range(0, N) if true_indicator[path, step] == 1), N - 1
        )
        val = excercise_option(stocks[path, step]["val"], K, op_type)
        estimator += math.exp(-r * (step + 1) * dt) * val
    return estimator / R

File: cpp/tilley/method.cc
#include <bits/stdc++.h>
#include <future>
#include <iostream>
#include <chrono>
#include "helper.h"
#include "sample.h"



struct Stock {
    double val;
    double current;
};


random_device rd;
mt19937 gen(rd());


double tilley(int P, int Q, double s0, double K, double sigma, double r, double T, int N, int op_type, bool sharp_boundary ) {
    int R = P * Q;
    double dt = T / N;
    double discount_rate = exp(-r * dt);
    vector<int> indexes(R);
    for (int i = 0; i < R; ++i)
        indexes[i] = i;
    vector<vector<Stock>> stocks(R, vector<Stock>(N));
    vector<vector<int>> true_indicator(R, vector<int>(N));
    normal_distribution<> d(0, sqrt(dt));
    for (int path = 0; path < R; ++path) {
        double prev = s0;
        for (int step = 0; step < N; ++step) {
            prev = generate_new_sample(prev, dt, sigma, r,gen,d);
            stocks[path][step].val = prev;
        }
        stocks[path][N - 1].current = excercise_option(prev, K, op_type);
    }

    for (int step = N - 2; step >= 0; --step) {
        auto key = [&](int x) { return stocks[x][step].val; };
        std::sort(indexes.begin(), indexes.end(), [&](int a, int b) { return op_type? key(a) < key(b) : key(a) > key(b); });
        vector<int> tentative_indicator(R, -1);
        vector<double> holding_values(Q);

        for (int bundle = 0; bundle < Q; ++bundle) {
            vector<double> bundle_currents;
            for (int path = 0; path < P; ++path) {
                int path_index = indexes[bundle * P + path];
                double current_value = stocks[path_index][step + 1].current;
                bundle_currents.push_back(current_value);
            }
            double holding_value = discount_rate * std::accumulate(bundle_currents.begin(), bundle_currents.end(), 0.0) / P;
            holding_values[bundle] = holding_value;
            for (int path = 0; path < P; ++path) {
                int path_index = indexes[bundle * P + path];
                double val = stocks[path_index][step].val;
                tentative_indicator[bundle * P + path] = (excercise_option(val, K, op_type) > holding_value) ? 1 : 0;
            }
        }

        int boundary = R;
        int i = R - 1;
        int max_zero_len = 0;
        while (i >= 0) {
            int j = i;
            if (tentative_indicator[i] == 1) {
                int one_len = 0;
                while (tentative_indicator[j] == 1 && j >= 0) {
                    ++one_len;
                    --j;
                }
                if (max_zero_len < one_len) {
                    boundary = j + 1;
                }
                i = j;
            } else {
                int zero_len = 0;
                while (tentative_indicator[j] == 0 && j >= 0) {
                    ++zero_len;
                    --j;
                }
                max_zero_len = max(zero_len, max_zero_len);
            }
            i = j;
        }

        vector<int> sharp_boundary_indicator(R);
        for (int i = boundary; i < R; ++i)
            sharp_boundary_indicator[i] = 1;

        vector<int>& indicator = sharp_boundary ? sharp_boundary_indicator : tentative_indicator;

        for (int bundle = 0; bundle < Q; ++bundle) {
            double holding_value = holding_values[bundle];
            for (int path = 0; path < P; ++path) {
                int path_index = indexes[bundle * P + path];
                double val = stocks[path_index][step].val;
                stocks[path_index][step].current = indicator[bundle * P + path] == 1 ? excercise_option(val, K, op_type) : holding_value;
            }
        }

        if (std::count(indicator.begin(), indicator.end(), 1) > 0) {
            int first_one = std::distance(indicator.begin(), std::find(indicator.begin(), indicator.end(), 1));
            true_indicator[indexes[first_one]][step] = 1;
            for (int j = step + 1; j < N; ++j)
                true_indicator[indexes[first_one]][j] = 0;
        }
    }

    double estimator = 0.0;
    for (int path = 0; path < R; ++path) {
        int step = std::find(true_indicator[path].begin(), true_indicator[path].end(), 1) - true_indicator[path].begin();
        step = (step == N) ? N - 1 : step;
        double val = excercise_option(stocks[path][step].val, K, op_type);
        estimator += exp(-r * (step + 1) * dt) * val;
    }
    return estimator / R;
}
